alphagov/tsuru-ansible
# [#98371938] Add Postgres standby server

Configure second postgres node as a standby and setup replication from master. Ensure playbook doesn't take too much time by running configuration of both master and standby in parallel. Provide easy way to promote master by defining master server in a variable.

Depends on https://github.com/alphagov/tsuru-terraform/pull/84
### How to test:

After doing ansible run, check that:
1 . initial backup exists, on master or slave do `sudo /usr/bin/envdir /etc/wal-e/ wal-e backup-list`, you should get st. like

```
wal_e.main   INFO     MSG: starting WAL-E
        DETAIL: The subcommand is "backup-list".
        STRUCTURED: time=2015-07-17T15:59:04.376411-00 pid=15043
name    last_modified   expanded_size_bytes wal_segment_backup_start    wal_segment_offset_backup_start wal_segment_backup_stop wal_segment_offset_backup_stop
base_000000010000000000000002_00000040  2015-07-16T14:38:52.484Z        000000010000000000000002    00000040
```

2 . Check postgres log on master: `sudo vim /var/log/postgresql/postgresql-9.3-main.log`, should contain st. like

```
wal_e.main   INFO     MSG: starting WAL-E
        DETAIL: The subcommand is "wal-push".
        STRUCTURED: time=2015-07-17T15:33:59.839938-00 pid=15056
wal_e.worker.upload INFO     MSG: begin archiving a file
        DETAIL: Uploading "pg_xlog/000000010000000000000006" to "s3://russ-mcp-postgres-backup/wal-e/wal_005/000000010000000000000006.lzo".
        STRUCTURED: time=2015-07-17T15:33:59.885894-00 pid=15056 action=push-wal key=s3://russ-mcp-postgres-backup/wal-e/wal_005/000000010000000000000006.lzo prefix=wal-e/ seg=000000010000000000000006 state=begin
wal_e.worker.upload INFO     MSG: completed archiving to a file
        DETAIL: Archiving to "s3://russ-mcp-postgres-backup/wal-e/wal_005/000000010000000000000006.lzo" complete at 61.661KiB/s.
        STRUCTURED: time=2015-07-17T15:34:01.308309-00 pid=15056 action=push-wal key=s3://russ-mcp-postgres-backup/wal-e/wal_005/000000010000000000000006.lzo prefix=wal-e/ rate=61.661 seg=000000010000000000000006 state=complete
wal_e.main   INFO     MSG: starting WAL-E
        DETAIL: The subcommand is "wal-push".
        STRUCTURED: time=2015-07-17T15:34:01.772151-00 pid=15086
wal_e.worker.upload INFO     MSG: begin archiving a file
        DETAIL: Uploading "pg_xlog/000000010000000000000007" to "s3://russ-mcp-postgres-backup/wal-e/wal_005/000000010000000000000007.lzo".
        STRUCTURED: time=2015-07-17T15:34:01.799727-00 pid=15086 action=push-wal key=s3://russ-mcp-postgres-backup/wal-e/wal_005/000000010000000000000007.lzo prefix=wal-e/ seg=000000010000000000000007 state=begin
wal_e.worker.upload INFO     MSG: completed archiving to a file
        DETAIL: Archiving to "s3://russ-mcp-postgres-backup/wal-e/wal_005/000000010000000000000007.lzo" complete at 1282.03KiB/s.
        STRUCTURED: time=2015-07-17T15:34:04.207910-00 pid=15086 action=push-wal key=s3://russ-mcp-postgres-backup/wal-e/wal_005/000000010000000000000007.lzo prefix=wal-e/ rate=1282.03 seg=000000010000000000000007 state=complete
wal_e.main   INFO     MSG: starting WAL-E
        DETAIL: The subcommand is "wal-push".
        STRUCTURED: time=2015-07-17T15:34:04.303460-00 pid=15094
wal_e.worker.upload INFO     MSG: begin archiving a file
        DETAIL: Uploading "pg_xlog/000000010000000000000007.00000028.backup" to "s3://russ-mcp-postgres-backup/wal-e/wal_005/000000010000000000000007.00000028.backup.lzo".
        STRUCTURED: time=2015-07-17T15:34:04.322704-00 pid=15094 action=push-wal key=s3://russ-mcp-postgres-backup/wal-e/wal_005/000000010000000000000007.00000028.backup.lzo prefix=wal-e/ seg=000000010000000000000007.00000028.backup state=begin
wal_e.worker.upload INFO     MSG: completed archiving to a file
        DETAIL: Archiving to "s3://russ-mcp-postgres-backup/wal-e/wal_005/000000010000000000000007.00000028.backup.lzo" complete at 00KiB/s.
        STRUCTURED: time=2015-07-17T15:34:04.945356-00 pid=15094 action=push-wal key=s3://russ-mcp-postgres-backup/wal-e/wal_005/000000010000000000000007.00000028.backup.lzo prefix=wal-e/ rate=00 seg=000000010000000000000007.00000028.backup state=complete
```

3 . Check postgres log on standby: `sudo vim /var/log/postgresql/postgresql-9.3-main.log`, should contain st. like

```
2015-07-16 14:41:26 UTC LOG:  entering standby mode
wal_e.main   INFO     MSG: starting WAL-E
        DETAIL: The subcommand is "wal-fetch".
        STRUCTURED: time=2015-07-16T14:41:26.956991-00 pid=28746
wal_e.operator.backup INFO     MSG: begin wal restore
        STRUCTURED: time=2015-07-16T14:41:26.981262-00 pid=28746 action=wal-fetch key=s3://russ-mcp-postgres-backup/wal-e/wal_005/000000010000000000000002.lzo prefix=wal-e/ seg=000000010000000000000002 state=begin
2015-07-16 14:41:27 UTC LOG:  incomplete startup packet
wal_e.blobstore.s3.s3_util INFO     MSG: completed download and decompression
        DETAIL: Downloaded and decompressed "s3://russ-mcp-postgres-backup/wal-e/wal_005/000000010000000000000002.lzo" to "pg_xlog/RECOVERYXLOG"
        STRUCTURED: time=2015-07-16T14:41:27.650980-00 pid=28746
wal_e.operator.backup INFO     MSG: complete wal restore
        STRUCTURED: time=2015-07-16T14:41:27.658997-00 pid=28746 action=wal-fetch key=s3://russ-mcp-postgres-backup/wal-e/wal_005/000000010000000000000002.lzo prefix=wal-e/ seg=000000010000000000000002 state=complete
2015-07-16 14:41:27 UTC LOG:  restored log file "000000010000000000000002" from archive
2015-07-16 14:41:27 UTC LOG:  redo starts at 0/2000028
2015-07-16 14:41:27 UTC LOG:  consistent recovery state reached at 0/20000F0
2015-07-16 14:41:27 UTC LOG:  database system is ready to accept read only connections
```

Also, `/var/lib/postgresql/9.3/main/recovery.conf` file should be present and pointing to master.
